### PR TITLE
Inomurko/vault+authority

### PR DESCRIPTION
**Pull request by InoMurko**

### CODE REVIEW

File: plasma_framework/truffle-config.js
Comment (boolafish) [inline plasma_framework/truffle-config.js:27]: I think this is misleading....you get deployed contract address from the file right (not private key...and there is no way to get a private key of the contract)?

This is one reason that is why I was suggesting a restructure of this comment: https://github.com/omgnetwork/plasma-contracts/pull/649#pullrequestreview-461371869

File: plasma_framework/migrations/1040_setup.js
Comment (boolafish) [inline plasma_framework/migrations/1040_setup.js:19]: A discussion on `VAULT`. Since we have removed the activate childchain function, so there is actually no "vault" interaction anymore, right? Could we name the flag with better naming that does not confuse. Or is the Multisig contract considered as a form of vault too?

**Comment by boolafish** [comment]
sorry not fully reviewed yet

File: plasma_framework/migrations/1040_setup.js
Comment (InoMurko) [reply plasma_framework/migrations/1040_setup.js:19]: The idea is that you'd use the vault *with* multisig maintainer account, so they're both enabled/disabled under one flag. Perhaps if there's need, we could also have that optional...

- So you could use the vault, but no multisig maintainer.
- You could use multisig maintainer, but not vault.

Would that be better?

File: plasma_framework/truffle-config.js
Comment (InoMurko) [reply plasma_framework/truffle-config.js:27]: yeah... I don't think this is used at all :D

Comment (InoMurko) [reply plasma_framework/truffle-config.js:27]: lemme review what you said in #649

File: plasma_framework/migrations/1040_setup.js
Comment (boolafish) [reply plasma_framework/migrations/1040_setup.js:19]: I am just wondering from the perspective of this repo, it does not make interaction to "vault" after we removed the activate childchain function. So it is not really a concept needed in this repo any more. So just wonder if there is better naming we can name this flag to...idk represent we are deploying with extra security care 😅

But yeah, I don't think we should complex this to 2 flag yet unless we see the need. I am just thinking if there can have a better naming for this concept but this comment could be minor. Not a blocker comment IMO

Comment (InoMurko) [reply plasma_framework/migrations/1040_setup.js:19]: Not sure I understand fully ... but Vault is still used and interacted with: https://github.com/omgnetwork/plasma-contracts/pull/678/files#diff-6ea48dfd3d7c2965be57aa58597eb191R14

because it needs to create the wallet and authority account that's later used.

Comment (boolafish) [reply plasma_framework/migrations/1040_setup.js:19]: oh! so actually we ... create the authority address via API during truffle migration?!
I misunderstood as we already have the authority account initiated/created beforehand.
I see

Comment (InoMurko) [reply plasma_framework/migrations/1040_setup.js:19]: we could do that but then ... it needs to be automated somewhere so I thought It might as well be here :)

File: plasma_framework/migrations/1050_output_results.js
Comment (souradeep-das) [inline plasma_framework/migrations/1050_output_results.js:18]: Maybe output doesn't show up for multisig?

File: .circleci/config.yml
Comment (InoMurko) [reply .circleci/config.yml:114]: I like em :D

File: plasma_framework/migrations/1050_output_results.js
Comment (InoMurko) [reply plasma_framework/migrations/1050_output_results.js:18]: yeah until the multisig actually executed all pending transactions we would have faulty outputs.

Comment (souradeep-das) [reply plasma_framework/migrations/1050_output_results.js:18]: Can we remove the vault check here? vault or no vault still just getting the deployed instances i mean 🤔

Comment (InoMurko) [reply plasma_framework/migrations/1050_output_results.js:18]: I think these three:
```
const ethVault = await plasmaFramework.vaults(config.registerKeys.vaultId.eth);
        const erc20Vault = await plasmaFramework.vaults(config.registerKeys.vaultId.erc20);
        const paymentExitGame = await plasmaFramework.exitGames(config.registerKeys.txTypes.payment);
```
would return faulty results since vaults aren't registered and exit games as well. is that what you meant?

Comment (souradeep-das) [reply plasma_framework/migrations/1050_output_results.js]: umm migration 1040 has register vault and exit game right?

Comment (souradeep-das) [reply plasma_framework/migrations/1050_output_results.js]: Maybe these values https://github.com/omgnetwork/plasma-contracts/pull/678/files#diff-a9446999f888af4dc205cab368614004R34

Comment (InoMurko) [reply plasma_framework/migrations/1050_output_results.js]:
```
1050_output_results.js
======================
{"authority_address":"0x7633fe8542c2218b5a25777477f63d395aa5afb4","eth_vault":"0x0000000000000000000000000000000000000000","erc20_vault":"0x0000000000000000000000000000000000000000","payment_exit_game":"0x0000000000000000000000000000000000000000","plasma_framework_tx_hash":"0x6fdc1e60ba970127239da7c3aa44121cd1a5f44e2a135434e686d44e38026c6b","plasma_framework":"0x2d697f6fb489326ecaabce7fc67678d3e8b838b1"}

- Saving migration to chain.
   > Saving migration to chain.
   -------------------------------------
   > Total cost:    
```
That's because https://github.com/omgnetwork/plasma-contracts/pull/678/files#diff-a9446999f888af4dc205cab368614004R54-R56 only puts the transactions into the multisig queue, not execute them. So techically they're waiting for the multisignature accounts to approve and call execute!

Comment (InoMurko) [reply plasma_framework/migrations/1050_output_results.js]: I mean yeah, I can leave it as is:
```
"eth_vault":"0x0000000000000000000000000000000000000000"
"erc20_vault":"0x0000000000000000000000000000000000000000"
"payment_exit_game":"0x0000000000000000000000000000000000000000"
```

Doesn't harm anyone 👍

Comment (souradeep-das) [reply plasma_framework/migrations/1050_output_results.js]: Aah i see. Let's see what we get here, if there is a problem we could probably just have whatever addresses are there logged? wdyt?

Comment (souradeep-das) [reply plasma_framework/migrations/1050_output_results.js]: awesome! we got what you said! 🚀

File: plasma_framework/truffle-config.js
Comment (InoMurko) [reply plasma_framework/truffle-config.js]: okay... https://github.com/omgnetwork/plasma-contracts/pull/649#pullrequestreview-461371869
to regards to your comment. The comment I think isn't really related to the issue at hand and I'm not sure how it's a better approach regardless. It seems (the current way) is a fairly standard way how people provide accounts to truffle migration. Having them in config doesn't really change anything much. Or am I not understanding something?

**Comment by InoMurko** [comment]
As a separate step I'll write a script that interacts with the gnosis multisig and executes all transactions.

File: MultiSigWalletOverride/truffle.js
Comment (souradeep-das) [inline MultiSigWalletOverride/truffle.js:27]: extra :D

File: plasma_framework/truffle-config.js
Comment (boolafish) [reply plasma_framework/truffle-config.js]: > and I'm not sure how it's a better approach regardless.

you do not need such `fs.readFileSync('../MultiSigWallet/build/multisig_instance')` thing (well...you do not need this line here too tbh) also all the [branching of getting vault address](https://github.com/omgnetwork/plasma-contracts/blob/bbdb558b4f61c54e34a7d8342c412ed45c13f3ec/plasma_framework/migrations/20_deploy_plasma_framework.js#L11-L17) could be hidden to a single place, reverting the dependency control of vault address to higher level of truffle migration script. In this case we have 2 options, `truffle-config.js` or `config.js`. I am not sure how to do that via `truffle-config.js` so I proposed `config.js`

> is a fairly standard way 

true, but current standard tooling (hd-wallet-provider) does not support what you need. That is what causing the complexity here.

hmmm...however, you have point me to a direction that it might be possible to just implement a new provider that would link with different way to "sending/approving transaction". Did a bit research and have not find enough documentation on this yet. So just throwing out idea at this moment. I will side track this direction differently and can/should be out of your effort of delivering this (unless it becomes obvious and it is super easy to do)

File: .circleci/config.yml
Comment (boolafish) [inline .circleci/config.yml:133]: I think DevOps has been asking for a docker to deploy things. We have a Dockerfile in the repo for this exact purpose and is tested on rinkeyby for each master merging (see the job `Deploy to Rinkeby`) though I don't think they have leveraged that as it is a bit new. Given our code would actually fail on random node version, I think it makes sense to support them with docker for a more stable deployer tool env setup.

Can we add docker support for them? So the ideal situation is this line is run inside a docker with the essential dependency installed already. Or we extend the existing `Deploy to Rinkeby` job with vault deployment.

File: MultiSigWalletOverride/migrations/2_deploy_contracts.js
Comment (boolafish) [inline MultiSigWalletOverride/migrations/2_deploy_contracts.js:23]: not sure I get what this means? Should the user retry with passing the account and confirmation using the data in this console.log? Also, might consider `console.error` instead
`Use: --accounts 0xasdf,0xfdsa --confirmations 2`

File: MultiSigWalletOverride/migrations/2_deploy_contracts.js
Comment (boolafish) [inline MultiSigWalletOverride/migrations/2_deploy_contracts.js:9]: probably you can use `indexOf` function here.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/indexOf

File: MultiSigWalletOverride/migrations/2_deploy_contracts.js
Comment (boolafish) [inline MultiSigWalletOverride/migrations/2_deploy_contracts.js:20]: minor: can we put `args[accountsIndex + 1].split(",")` into a variable as this is used afterward again to be more readable? same as the confirmation one.

File: MultiSigWalletOverride/migrations/2_deploy_contracts.js
Comment (InoMurko) [reply MultiSigWalletOverride/migrations/2_deploy_contracts.js:23]: It just explains how the migrations should be run, because I've overriden the way gnosis does it. Ill use console.error, tnx!

File: .circleci/config.yml
Comment (InoMurko) [reply .circleci/config.yml:133]: yeah we can def. do that. I'll do that in the next PR so that I don't make this too big.

File: plasma_framework/migrations/1041_setup_vault.js
Comment (souradeep-das) [inline plasma_framework/migrations/1041_setup_vault.js:69]: looks like we wait for the receipt but it isn't logged at the end 🤔 Does anything silently fail?

**Comment by souradeep-das** [comment]
Should we also enable the `Deploy to Rinkeby` workflow?

File: plasma_framework/migrations/1041_setup_vault.js
Comment (InoMurko) [reply plasma_framework/migrations/1041_setup_vault.js:69]: I'm not sure I understand. What isn't logged? Ah, do you think I should look at status from the transactionReceipt and abort if it's not true?

Comment (souradeep-das) [reply plasma_framework/migrations/1041_setup_vault.js:69]: Ohh the transactionReceipt body's aren't logged i think. Yeah I was also thinking about looking at the status and handling if a failed transaction, though not a necessity maybe but might be good to have

Comment (souradeep-das) [reply plasma_framework/migrations/1041_setup_vault.js:69]: Probably just logging out transaction receipts to show which failed or aborting

Comment (InoMurko) [reply plasma_framework/migrations/1041_setup_vault.js:69]: I think I'll abort. Because we insert 8 transactions into the gnosis multisig, all of them need to be executed "in order" they were submitted. So it's easier to know where things broke and where to continue.

**Comment by InoMurko** [comment]
> Should we also enable the `Deploy to Rinkeby` workflow?

So that we deploy twice? Once with vault and the second time without?

**Comment by souradeep-das** [comment]
yeah 👍

**Comment by InoMurko** [comment]
I've just noticed `Deploy to Rinkeby` uses a docker approach. So I'll do that in the subsequent PR where I actually install all dependencies that we need to support vault and gnosis stuff.

File: .gitignore
Comment (boolafish) [inline .gitignore:54]: really minor: GH dislike file without extra line 🤷‍♂️ always show up on review page

File: plasma_framework/migrations/1041_setup_vault.js
Comment (boolafish) [inline plasma_framework/migrations/1041_setup_vault.js:69]: probably at least add a log or print sth before sleeping so we would know what we are waiting and the code is still running if it gets slow....same for other places

File: plasma_framework/migrations/1041_setup_vault.js
Comment (boolafish) [inline plasma_framework/migrations/1041_setup_vault.js:61]: omg, it is so great it is not callback now compare to the first version 🙇‍♂️ it was hard

File: plasma_framework/migrations/1041_setup_vault.js
Comment (boolafish) [inline plasma_framework/migrations/1041_setup_vault.js:64]: I believe js prefer `===` instead of `==` for stricter comparison though I forgot what exactly the difference is 😅 But in general it is better to forgot `==` exists.

File: plasma_framework/migrations/1041_setup_vault.js
Comment (boolafish) [inline plasma_framework/migrations/1041_setup_vault.js:73]: minor: I would appreciate if you could break these steps into smaller functions and call it here for better readability

File: plasma_framework/migrations/1041_setup_vault.js
Comment (boolafish) [inline plasma_framework/migrations/1041_setup_vault.js:81]: same here `!==` is preferred. Apply to all the  other `if` inside `while` I think

File: MultiSigWalletOverride/migrations/2_deploy_contracts.js
Comment (kevsul) [inline MultiSigWalletOverride/migrations/2_deploy_contracts.js:19]: Minor: Using `await` here might be better / easier to read.
As it is, if the deploy fails you'll still get a `Wallet deployed` log, which could be confusing.

Either use `await` or move the `console.log("Wallet deployed");` into the `then` block

File: plasma_framework/migrations/1041_setup_vault.js
Comment (kevsul) [inline plasma_framework/migrations/1041_setup_vault.js:63]: Minor. `buildDir` doesn't seem the right name for this variable.

File: plasma_framework/migrations/1041_setup_vault.js
Comment (kevsul) [inline plasma_framework/migrations/1041_setup_vault.js:19]: expectedBlockTime is one second?

File: plasma_framework/migrations/1041_setup_vault.js
Comment (kevsul) [reply plasma_framework/migrations/1041_setup_vault.js:80]: I agree. There seems to be quite a bit of duplicated code that could be extracted

File: plasma_framework/migrations/1041_setup_vault.js
Comment (InoMurko) [reply plasma_framework/migrations/1041_setup_vault.js:19]: it's more like how frequently we're willing to pull for results

**Review by kevsul — APPROVED** [review]

**Pull request merged**
